src/multi.c

```diff
@@ -348,17 +351,43 @@ void touchWatchedKey(redisDb *db, robj *key) {
     clients = dictFetchValue(db->watched_keys, key);
     if (!clients) return;
 
+    int key_exists = dictFind(db->dict, key->ptr) != NULL;
+
     /* Mark all the clients watching this key as CLIENT_DIRTY_CAS */
     /* Check if we are already watching for this key */
     listRewind(clients,&li);
     while((ln = listNext(&li))) {
         client *c = listNodeValue(ln);
 
+        if (!key_exists) {
+            /* Was the key expired already when the client watched it? If so,
+             * there is no logical change regarding the key's existence. */
+            listIter wk_li;
+            listNode *wk_ln;
+            listRewind(c->watched_keys, &wk_li);
+            while ((wk_ln = listNext(&wk_li))) {
+                watchedKey *wk = listNodeValue(wk_ln);
+                if (wk->db == c->db && equalStringObjects(key, wk->key)) {
+                    /* Found the key. */
+                    if (wk->key->lru == 1) {
+                        /* The key was already expired when WATCH was called.
+                         * Now the key is explicitly deleted. */
+                        wk->key->lru = 0;
+                        goto skip_client;
```

Comment:
i have a concern about this extra loop. (even if only when a key is missing from the db).
actually i may also be concerned about the extra dictFind.

maybe we should modify the data we put in `db->watched_keys` per client to contains a struct with a list node to `c->watched_keys`, rather than just a client struct pointer?

Reply:
Do you mean turn the values in values `db->watched_keys` into lists of `struct { client *c; listNode *client_wk_node; }`?

* This is an extra struct that needs to be allocated per client and key, apart from the struct watchedKey.
* We need to access the client from here too (for marking it as dirty when a key changes) so we need a client pointer too.

What if instead turn `db->watched_keys` into a list of `watchedKey` pointers and add a client pointer inside the watchedKey struct? This struct is already allocated per client and key.

If looping over `c->watched_keys` is a concern, we can turn this list into a dict. However, this list is usually very short (1-2) so I think there is no benefit. A dict probably has more overhead than a list for a list in this case.

Reply:
i think this list can be very long in some cases.
your suggestion about storing a pointer to the `watchedKey` struct and have it hold a pointer to the client seem ok.
maybe @soloestoy can have another suggestion?

Reply:
> i think this list can be very long in some cases.

List vs dict overhead for 1 element:

* dict (4 words) + hash table (dictEntry* array, min size 4 words) + dictEntry (4 words) = 16 words
* list (8 words) + listNode (4 words) = 12 words

If we want to use dicts here, perhaps we could optimize dictFind for very small dicts so that we don't hash the key, let's say for dicts of size == 1 or size < 4. Then we just try all keys. This might speed up HGET too in some cases.

> your suggestion about storing a pointer to the watchedKey struct and have it hold a pointer to the client seem ok.

OK, I'll try this first.